# Mandatory Exercise - Session 2

### Students: Nafis Banirazi & Jan Carbonell

### Lab Objective:
The Objective of this lab are the following: 
- To read all pairs of sentences from a trial set, compute their similarities using Jaccard distance. 
- Compare the results with the gold standard by giving the Pearson correlation.


In [1]:
# initial set of imports
import nltk
from nltk.metrics import jaccard_distance
from scipy.stats import pearsonr

#value initialization and instantiation
d = {}
tests = []
standard = []

We now proceed to open and read the input file

In [2]:
for line in open('./trial/STS.input.txt','r'):
    line = line.strip().split("\t")
    if line[0] in d:
        pass
    else:
        d[line[0]] = line[1:] 

We compute the Jaccard distance. In this step, we must first tokenize the sentences. We then take the **sets**; *unique values of those tokenized sentences*, and compute the **jaccard similarity as 1 - jaccard distance**. 

In [5]:
for key in d:
    sents = [nltk.word_tokenize(s) for s in d[key]]
    print(sents)
        
    w1 = set(sents[0])
    w2 = set(sents[1])

    # jaccard similarity 1 - jaccard distance
    dist = jaccard_distance(w1, w2)
    jaccard_similarity = 1 - dist
    tests.append(round(jaccard_similarity,3))
print(tests)

[['The', 'bird', 'is', 'bathing', 'in', 'the', 'sink', '.'], ['Birdie', 'is', 'washing', 'itself', 'in', 'the', 'water', 'basin', '.']]
[['In', 'May', '2010', ',', 'the', 'troops', 'attempted', 'to', 'invade', 'Kabul', '.'], ['The', 'US', 'army', 'invaded', 'Kabul', 'on', 'May', '7th', 'last', 'year', ',', '2010', '.']]
[['John', 'said', 'he', 'is', 'considered', 'a', 'witness', 'but', 'not', 'a', 'suspect', '.'], ['``', 'He', 'is', 'not', 'a', 'suspect', 'anymore', '.', "''", 'John', 'said', '.']]
[['They', 'flew', 'out', 'of', 'the', 'nest', 'in', 'groups', '.'], ['They', 'flew', 'into', 'the', 'nest', 'together', '.']]
[['The', 'woman', 'is', 'playing', 'the', 'violin', '.'], ['The', 'young', 'lady', 'enjoys', 'listening', 'to', 'the', 'guitar', '.']]
[['John', 'went', 'horse', 'back', 'riding', 'at', 'dawn', 'with', 'a', 'whole', 'group', 'of', 'friends', '.'], ['Sunrise', 'at', 'dawn', 'is', 'a', 'magnificent', 'view', 'to', 'take', 'in', 'if', 'you', 'wake', 'up', 'early', 'enoug

And now, we open the Golden Standard file and calculate the perason correlation with and without lemmatization. 

In [4]:
for line in open('./trial/STS.gs.txt','r'):
    line = line.strip().split("\t")
    standard.append(int(line[1]))

a = pearsonr(standard[0:6], tests[0:6])[0]
print('Pearson correlation:', round(a,3))

Pearson correlation: 0.396


## Conclusions

Over the development of this lab, we have computed the similarities between different pairs of sentences; calculated their jaccard distance.

We have also compared them with the gold standard and have achieved the following results

*Pearson correlation: 0.396*